mycore-restapi/src/main/java/org/mycore/restapi/v2/MCRRestDerivateContents.java

```diff
@@ -0,0 +1,547 @@
+            LogManager.getLogger().info("Creating directory: {}", mcrPath);
+            try {
+                doWithinTransaction(() -> Files.createDirectory(mcrPath));
+            } catch (IOException e2) {
+                throw new InternalServerErrorException(e2);
+            }
+            return Response.status(Response.Status.CREATED).build();
+        }
+    }
+
+    private static void doWithinTransaction(IOOperation op) throws IOException {
+        MCRSession mcrSession = MCRSessionMgr.getCurrentSession();
+        try {
+            mcrSession.beginTransaction();
+            op.run();
+        } finally {
+            if (mcrSession.transactionRequiresRollback()) {
+                mcrSession.rollbackTransaction();
+            } else {
+                mcrSession.commitTransaction();
+            }
+        }
+    }
+
+    private Response updateOrCreateFile(InputStream contents, MCRPath mcrPath) {
+        MCRFileAttributes fileAttributes;
+        try {
+            fileAttributes = Files.readAttributes(mcrPath, MCRFileAttributes.class);
+            if (!fileAttributes.isRegularFile()) {
+                throw new BadRequestException("Overwrite directory with file: " + mcrPath);
+            }
+        } catch (IOException e) {
+            //does not exist
+            LogManager.getLogger().info("Creating file: {}", mcrPath);
+            try {
+                OutputStream out = Files.newOutputStream(mcrPath, StandardOpenOption.CREATE_NEW);
+                try {
+                    IOUtils.copy(contents, out, BUFFER_SIZE);
+                } finally {
+                    //close writes data to database
+                    doWithinTransaction(out::close);
+                }
+            } catch (IOException e2) {
+                try {
+                    doWithinTransaction(() -> Files.deleteIfExists(mcrPath));
+                } catch (IOException e3) {
+                    LogManager.getLogger().warn("Error while deleting incomplete file.", e3);
+                }
+                throw new InternalServerErrorException(e2);
+            }
+            return Response.status(Response.Status.CREATED).build();
+        }
+        //file does already exist
+        Date lastModified = new Date(fileAttributes.lastModifiedTime().toMillis());
+        EntityTag eTag = getETag(fileAttributes);
+        Optional<Response> cachedResponse = MCRRestUtils.getCachedResponse(request.getRequest(), lastModified,
+            eTag);
+        if (cachedResponse.isPresent()) {
+            return cachedResponse.get();
+        }
+        LogManager.getLogger().info("Updating file: {}", mcrPath);
+        int memBuf = MCRConfiguration2.getOrThrow("MCR.FileUpload.MemoryThreshold", Integer::parseInt);
+        java.io.File uploadDirectory = MCRConfiguration2.getOrThrow("MCR.FileUpload.TempStoragePath",
+            java.io.File::new);
+        try (DeferredFileOutputStream dfos = new DeferredFileOutputStream(memBuf, mcrPath.getOwner(),
```

Comment:
To decrease the complexity i would suggest to always write to a temp file. You will also have less problems if the request will be aborted.

Reply:
abortion is handled in the code and for new big files the code is faster